### DISCUSSION TITLE
Run Python Scripts from Function keys

### DISCUSSION THREAD

**Discussion by roehligas** [cwapi3d/cwapi3dpython#146] [Ideas]
I have sent this wish in several times in support, so i just add it in here for formality as well.

The user experience would be increased extreme if the user could launch Python scripts from the function keys.
Right now the amount of quickly accessible scripts is limited by the space for buttons inside the window.

Another aspect is the fact that you save a lot of time by not having to move the mouse to the buttons, it really interrupts the user performance when you have tasks with high repetition.

**Comment by LPeccoud** [comment]
Good morning,
Here is my method:
Assign to a function key (menu Help /Help Functions F1-F12) : Category = Extra, Command = External Program.
After support on the function key, will appear in the right menu, the list of your scripts under API.x64.
Tap the first letter of the name of the script, then "ENTER".
The script starts ....

![Capture d’écran 2024-03-09 125430](https://github.com/cwapi3d/cwapi3dpython/assets/163459469/e8158f85-57bf-4cfe-99f8-1d49613e248a)
![Capture d’écran 2024-03-09 125624](https://github.com/cwapi3d/cwapi3dpython/assets/163459469/03ac8dbb-c021-4584-b65c-3a8ad6b2515f)

**Reply by roehligas** [reply]
Hi,

thanks a lot for the input!!!! 
That's a neat function, this helps quite well!

However, in my mind it still requires two unnecessary key-presses.
I'll see if we can build upon this with some makro-function to make it quick. :)

**Comment by roehligas** [comment]
Could a low-effort high-reward solution be to create 10-48 folders somewhere in the userprofile similar to API.x64 where each folder represents a function that can be called by function-key?

At the end of the day there's a limit to how many function keys are available (48 as far as i counted)...

**Comment by Brunner246** [comment]
you could create a Ui and then delegate a script execution as required via Buttons or shortcuts (https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QShortcut.html). 

There is `def run_external_program_from_custom_directory(file_path: str) -> bool` in utility controller.

**Reply by roehligas** [reply]
PyQt only works when the script-window is active, i tested.

```
import sys
from PyQt5.QtWidgets import QApplication, QMainWindow, QLabel, QShortcut
from PyQt5.QtGui import QKeySequence

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()

        self.setWindowTitle("Shortcut Example")
        self.setGeometry(100, 100, 400, 200)

        self.label = QLabel("Press Ctrl+G to greet!", self)
        self.label.setGeometry(50, 50, 300, 50)

        # Create a shortcut for greeting
        shortcut = QShortcut(QKeySequence("Ctrl+G"), self)
        shortcut.activated.connect(self.greet)

    def greet(self):
        self.label.setText("Hello!")

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())
```

We also have way too many scripts for the current system. When i add all 30 Scripts/Fucntions that I want to have accessible the names in the window get so small that you can't read them anymore.

![image](https://github.com/cwapi3d/cwapi3dpython/assets/66385500/a98865af-7a90-4b0a-935b-b4a50a77094b)

I've looked into a program named autohotkey and it looks very promising. I used the code below in a AutoHotkey-script to put a sequence of buttons on Ctrl+F5. I then put the external program command on F7. So when i press Ctrl+F5, the external program command is fired and then 10 milliseconds later autohotkey presses the "c" button twice to jump the cursor to the second script starting with "c" and then presses Enter to run the script.

It works, but I really recommend creating a simpler approach, because as mentioned above having so many scripts in the APIx64 folder makes the text of all scripts unreadable....

```
^F5:: ; 
    Send, {F7} ; 
    Sleep, 10 ; 
    Send, c ; 
    Send, c ; 
    Send, {Enter} ; 
return
```


Update:
I forgot to mention that AutoHotkey can be set up so that i checks in what window you currently are. So it doesn't affect other windows.

**Comment by usai-sylvain** [comment]
Continuing the streak of digging old issues... 🧟 

I've heard similar requests multiple times so I gave it a try. The following python script works with the Function key shortcut by @LPeccoud, I create a plugin whose name starts with "___" so I'm sure it will stay on top of the list.

```python
import os
import sys
import importlib
from pathlib import Path

import utility_controller as uc


def main():
    # Get the API.x64 base directory
    api_base = Path(__file__).parent.parent

    # get all folder paths in the API.x64 base directory
    folder_paths = [f.path for f in os.scandir(api_base) if f.is_dir()]
    # sort them alphabetically
    folder_paths.sort()

    # get the folder name from the folder path
    folder_names = [Path(f).name for f in folder_paths]

    prompt_parts = []
    for i, folder_name in enumerate(folder_names):
        prompt_parts.append(f"{i}={folder_name}")
    prompt = ", ".join(prompt_parts)

    

    # get the user input
    user_input = uc.get_user_int(prompt)

    # get the folder name from the user input
    for extension in [".py", ".dll"]:
        plugin_path = folder_paths[int(user_input)] + "\\" + folder_names[int(user_input)] + extension
        if Path(plugin_path).exists():
            print(f"Opening {plugin_path}...")
            result = uc.run_external_program_from_custom_directory(plugin_path)
            print(result)
            break
```

The issue is that you still need to type a number and press enter... which mean you would be faster using key down from the list you get from the Fn shortcut : 
<img width="430" height="512" alt="image" src="https://github.com/user-attachments/assets/c0ed74c1-2a24-4a5d-857d-ad7ab9ec534a" />

One way around would be to use `mc.display_simple_menu(menu_options)` and have the option to define shortcut for each of the menu_options like native cadwork menu have : 
<img width="321" height="635" alt="image" src="https://github.com/user-attachments/assets/621bab46-56e0-4d61-aac7-6fde133a444c" />

@Brunner246 does the "define shortcut for each of the menu_options" request make sense ?
